DEP: Add back `ndarray.__[sg]etslice__`, but deprecate it

eric-wieser opened the pull request:
This is added in a way that means that CPython will not call these for
normal indexing.

This partially undoes #8592, and is a competing approach to #8943.


Other removals in #8592:

* `np.r_.__getslice__` - `np.r_` is pretty broken anway (#4559 #8815 #8818), and it seems really unlikely that anyone would actually call this
* `np.ma.__getslice__` - this falls through to `ndarray.__getslice__` just fine now
* `np.nditer.__getslice__` - seems incredibly unlikely someone would subclass this

Based on #8887, it's possible this causes a problem on pypy (in that `a[1:2]` causes a deprecation warning) - can anyone test this?

eric-wieser committed:
DEP: Add back ndarray.__[sg]etslice__

These now delegate to __getitem__, so subclasses no longer need to define them

charris commented:
Looks like indexing ends up calling {get,set}slice by default if they are there. I'd just remove the deprecation and wait until we drop 2.7. The main point here is the reuse the code for {get,set} item. After that, looks like one test failure will remain.

eric-wieser commented:
@charris: Am I right in thinking that it makes sense to put this in `__dict__` but not `tp_mapping`? My thoughts are that this avoids an extra roundtrip from `PyObject*` to `int` and back,

charris commented:
I think you want this in `numpy/testing/nosetester.py` ~line 423.
```
+                sup.filter(DeprecationWarning, message="in 3\.x, __setslice__")
+                sup.filter(DeprecationWarning, message="in 3\.x, __getslice__")
```

I suspect you removed that yourself ;)

charris commented:
> Am I right in thinking that it makes sense to put this in __dict__ but not tp_mapping? 

I think that should work...

charris commented:
Just needs the nosetester fix and I think it is ready to go.

juliantaylor commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -2417,6 +2417,44 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return c;
 }
 
+static PyObject *
```
these need to be in an `ifndef NPY_PY3K`

eric-wieser committed:
Merge remote-tracking branch 'upstream/master' into remove-setslice

eric-wieser commented:
@charris: Done, and with a more accurate commit message

charris commented:
Merged. Thanks Eric.

juliantaylor commented:
please rebase instead of merging master back in for these small branches.

eric-wieser commented:
The merge here was deliberate - I wanted the parent of this commit to be the previous PR, so that it was easy to check what the sum total of changes was.

charris commented:
Wouldn't a rebase have done as well? What am I missing?

eric-wieser commented:
@charris: If I rebased, there would be no way to ask for a diff between this change and the one that it undoes, without getting everything in the middle (unless I rebased the already-merged PR, which would be weird, and create duplicate commits)

charris commented:
Are you sure? What you would see would be the difference from current master head, which would look, IIANM, exactly like this does except with a cleaner history.

eric-wieser commented:
Nothing. The difference would be what you see as a diff between f031cdf83593907431ca2666d8484cb1c36315bd (the last non-merge commit in this PR) and 98b3127fa61e09387e1f7cf925acfcb36043e48a (the commit before __getslice__ was removed)
